### PR TITLE
[WIP] Add SeenReplyCount column to ThreadMemberships

### PR DESCRIPTION
**Pull request by ashishbhate**
#### Release Note

```release-note

```

### CODE REVIEW

File: app/channel.go [hunk 1]
```diff
@@ -2724,11 +2724,15 @@ func (a *App) markChannelAsUnreadFromPostCRTUnsupported(postID string, userID st
 		}
 		// If threadmembership already exists but user had previously unfollowed the thread, then follow the thread again.
 		threadMembership.Following = true
-		threadMembership.LastViewed = post.UpdateAt - 1
-		threadMembership.UnreadMentions, err = a.countThreadMentions(user, rootPost, channel.TeamId, post.UpdateAt-1)
+		threadMembership.LastViewed = post.CreateAt - 1
+		threadMembership.UnreadMentions, err = a.countThreadMentions(user, rootPost, channel.TeamId, post.CreateAt-1)
```
Comment (ashishbhate) [inline app/channel.go:2728]: Related fix:
These should be CreateAt, not UpdateAt. The subsequent methods also filter posts / count messages by CreateAt, so using UpdateAt will lead to less than the expected number of Mentions/Unreads under some circumstances.

Needs tests.

File: store/sqlstore/post_store.go [hunk 2]
```diff
@@ -2620,64 +2620,78 @@ func (s *SqlPostStore) deleteThread(transaction *sqlxTxWrapper, postId string, d
...
+		return errors.Wrap(err, "failed to create SQL query to update thread memberships")
 
-		if err != nil {
-			return errors.Wrap(err, "failed to update Threads")
-		}
 	}
-	return nil
+
+	_, err = transaction.Exec(updateMembershipsQueryString, args...)
+	return errors.Wrap(err, "failed to update ThreadMemberships")
```
Comment (ashishbhate) [inline store/sqlstore/post_store.go:2693]: On Post delete, we don't update the `MsgCount`s in the channels, so we don't have to update channelMemberships. 

We update the ReplyCount for Threads (to show the correct reply count in the UI), so we have to update ThreadMemberships as well.

File: db/migrations/postgres/000082_threadmemberships_seenreplycount.up.sql [hunk 3]
```diff
@@ -0,0 +1,17 @@
+ALTER TABLE threadmemberships ADD COLUMN IF NOT EXISTS seenreplycount bigint DEFAULT 0;
```
Comment (ashishbhate) [inline db/migrations/postgres/000082_threadmemberships_seenreplycount.up.sql:1]: In my local DB I found some ThreadMemberships with no corresponding Threads. I'm not sure how this could have happened, but I've added the `Default 0` to guard against subsequent query failures if seenreplycount remains `NULL`

File: store/storetest/thread_store.go [hunk 4]
```diff
@@ -349,24 +349,6 @@ func testThreadStorePopulation(t *testing.T, ss store.Store) {
...
-	t.Run("Empty participantID should not appear in thread response", func(t *testing.T) {
-		newPosts := makeSomePosts()
-		opts := store.ThreadMembershipOpts{
-			Following:             true,
-			IncrementMentions:     false,
-			UpdateFollowing:       true,
-			UpdateViewedTimestamp: false,
-			UpdateParticipants:    true,
-		}
-		m, err := ss.Thread().MaintainMembership("", newPosts[0].Id, opts)
```
Comment (ashishbhate) [inline store/storetest/thread_store.go:361]: With the the addition of the SeenReplyCount column, the `UserID` argument to MaintainMembership can't be empty (PostId being empty doesn't make sense either). This was never actually supported, and it seems to be check we missed previously. I've added a check to the MaintainMembership function and removed this test.

File: store/storetest/thread_store.go [hunk 5]
```diff
@@ -332,8 +332,8 @@ func testThreadStorePopulation(t *testing.T, ss store.Store) {
 		require.NoError(t, err)
 		require.Equal(t, int64(2), th.UnreadReplies)
 
-		m.LastViewed = newPosts[2].UpdateAt + 1
-		_, err = ss.Thread().UpdateMembership(m)
+		// m.LastViewed = newPosts[2].UpdateAt + 1
```
Comment (ashishbhate) [inline store/storetest/thread_store.go:335]: Will remove with next round of fixes

File: app/channel.go [hunk 6]
```diff
@@ -2724,11 +2724,15 @@ func (a *App) markChannelAsUnreadFromPostCRTUnsupported(postID string, userID st
...
 		if err != nil {
 			return nil, err
 		}
+		threadMembership.SeenReplyCount, sErr = a.Srv().Store.Thread().GetReplyCountUntil(rootPost.Id, post.CreateAt-1)
```
Comment (ashishbhate) [inline app/channel.go:2732]: Marking as unread leads to this extra store which scans the posts table.

File: store/sqlstore/thread_store.go [hunk 7]
```diff
@@ -677,29 +667,59 @@ func (s *SqlThreadStore) MarkAllAsReadByTeam(userId, teamId string) error {
...
+	defer finalizeTransactionX(transaction)
+
+	queryString, args, err := s.getQueryBuilder().
+		Select("COUNT(Posts.Id)").
+		From("Posts").
+		Where(sq.And{
+			sq.Eq{"Posts.RootId": threadId},
+			sq.Eq{"Posts.DeleteAt": 0},
+			sq.LtOrEq{"Posts.CreateAt": timestamp},
+		}).
```
Comment (ashishbhate) [inline store/sqlstore/thread_store.go:683]: This method does dual duty:

if  the `timestamp` argument is greater than last reply at of the thread, then it marks the thread as read.

If the `timestamp` argument is less that the last reply at of the thread, then it marks the thread as unread after that timestamp.


In the first case, we shouldn't have to scan the posts table, we can simply assign `ThreadMemberships.SeenReplyCount = Threads.ReplyCount`. 
The client can't tell us if its seen the whole thread, because it may have out-of-date information about the thread, i.e. it may not have the latest `LastReplyAt` of the Thread. 

We could check if the `timestamp` is after the `LastReplyAt` of the thread. In this case we set `ThreadMemberships.SeenReplyCount = Threads.ReplyCount`. In the complementary case, we do the posts table scan as above. 

I'm still investigating if this extra check is going to give us a performance win over always scanning the posts table.

**Review by ashishbhate — COMMENTED** [review]
The PR is ready for the first pass of reviews. The PR still needs some polish and new tests to validate some edge cases. 

I'd like to do some performance testing first to check if the changes in the PR are worth the effort.

File: store/sqlstore/thread_store.go [hunk 8]
```diff
@@ -819,12 +840,40 @@ func (s *SqlThreadStore) DeleteMembershipForUser(userId string, postId string) e
 // - channel marked unread
 // - user explicitly following a thread
 func (s *SqlThreadStore) MaintainMembership(userId, postId string, opts store.ThreadMembershipOpts) (*model.ThreadMembership, error) {
+	if userId == "" {
+		return nil, errors.New("userId must be specified for thread membership")
+	}
+	if postId == "" {
+		return nil, errors.New("threadId must be specified for thread membership")
+	}
```
Comment (agnivade) [inline store/sqlstore/thread_store.go:848]: Any reason to add these? Can these be enforced in the app layer?

**Review by agnivade — COMMENTED** [review]
Gave it a first pass. Looks very good. Looking forward to getting the tests added.

**Comment by ashishbhate** [comment]
@agnivade @lieut-data Since we've hit our performance goals, do you think this PR is worth finishing?

Perhaps in a much later release? We can let things settle down for now, with fewer moving parts during and right after GA.

**Comment by lieut-data** [comment]
@ashishbhate, I'm definitely in favour of stabilizing things for GA given where we're at now, and keeping this in our back pocket.

**Comment by agnivade** [comment]
I definitely think this is something we should merge, but agree with Jesse that we should let things stabilize for GA first.

File: db/migrations/mysql/000082_threadmemberships_seenreplycount.up.sql [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+));
+
+PREPARE addColumnIfNotExists FROM @preparedStatement;
+EXECUTE addColumnIfNotExists;
+DEALLOCATE PREPARE addColumnIfNotExists;
+
+UPDATE
+	ThreadMemberships, Threads
+SET
+	SeenReplyCount = Threads.ReplyCount WHERE ThreadMemberships.LastViewed >= Threads.LastreplyAt AND ThreadMemberships.PostId = Threads.PostId;
```
Comment (lieut-data) [inline db/migrations/mysql/000082_threadmemberships_seenreplycount.up.sql:19]: nit, not sure of the exact SQL style we follow, but can we pull `WHERE` down at least?
```suggestion
	SeenReplyCount = Threads.ReplyCount 
WHERE 
	ThreadMemberships.LastViewed >= Threads.LastReplyAt AND
	ThreadMemberships.PostId = Threads.PostId;
```

File: db/migrations/mysql/000082_threadmemberships_seenreplycount.up.sql [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+
+UPDATE
+	ThreadMemberships, Threads
+SET
+	SeenReplyCount = (SELECT COUNT(Posts.Id) FROM Posts WHERE Posts.DeleteAt = 0 AND Posts.RootId = ThreadMemberships.PostId AND Posts.CreateAt < ThreadMemberships.LastViewed)
```
Comment (lieut-data) [inline db/migrations/mysql/000082_threadmemberships_seenreplycount.up.sql:24]: Should the condition be less-than-or-equal, e.g. `Posts.CreateAt <= ThreadMemberships.LastViewed`?

File: db/migrations/mysql/000082_threadmemberships_seenreplycount.up.sql (hunk 9 above)
Comment (lieut-data) [reply db/migrations/mysql/000082_threadmemberships_seenreplycount.up.sql:19]: Also, not `Threads.LastReplyAt` casing above.

File: db/migrations/postgres/000082_threadmemberships_seenreplycount.up.sql [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+UPDATE
+	threadmemberships
+SET
+	seenreplycount = threads.replycount from threads where threadmemberships.lastviewed >= threads.lastreplyat AND threadmemberships.postid = threads.postid;
```
Comment (lieut-data) [inline db/migrations/postgres/000082_threadmemberships_seenreplycount.up.sql:6]:
```suggestion
	seenreplycount = threads.replycount from threads 
WHERE
	threadmemberships.lastviewed >= threads.lastreplyat AND 
	threadmemberships.postid = threads.postid;
```

File: db/migrations/postgres/000082_threadmemberships_seenreplycount.up.sql [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+
+UPDATE
+	threadmemberships
+SET
+	seenreplycount = (SELECT COUNT(posts.Id) FROM posts WHERE posts.deleteat = 0 AND posts.rootid = threadmemberships.postid AND posts.createat < threadmemberships.lastviewed)
```
Comment (lieut-data) [inline db/migrations/postgres/000082_threadmemberships_seenreplycount.up.sql:11]: As above, should this be `posts.createat <= threadmemberships.lastviewed`?

Comment (lieut-data) [reply db/migrations/postgres/000082_threadmemberships_seenreplycount.up.sql:11]: Also, nit: `posts.Id`, whereas we lowercase everything else.

File: store/sqlstore/thread_store.go [hunk 13]
```diff
@@ -494,21 +480,10 @@ func (s *SqlThreadStore) GetThreadForUser(teamId string, threadMembership *model
...
-			sq.Eq{"Posts.DeleteAt": 0},
-		}).ToSql()
-	if err != nil {
-		return nil, errors.Wrapf(err, "failed to build subquery to count unread replies for getting thread for user id=%s, post id=%s", threadMembership.UserId, threadMembership.PostId)
-	}
-
 	fetchConditions := sq.And{
 		sq.Or{sq.Eq{"Channels.TeamId": teamId}, sq.Eq{"Channels.TeamId": ""}},
 		sq.Eq{"Threads.PostId": threadMembership.PostId},
+		sq.Eq{"ThreadMemberships.UserId": threadMembership.UserId},
```
Comment (lieut-data) [inline store/sqlstore/thread_store.go:486]: Should we inline `fetchConditions` now that it seems to be used only once?

File: store/sqlstore/thread_store.go [hunk 14]
```diff
@@ -624,13 +599,21 @@ func (s *SqlThreadStore) MarkAllAsReadByChannels(userID string, channelIDs []str
...
+
+	} else {
+		querySQL = s.getQueryBuilder().Update("ThreadMemberships", "Threads")
+	}
+
+	query, args, err := querySQL.
 		Set("LastViewed", timestamp).
 		Set("UnreadMentions", 0).
 		Set("LastUpdated", model.GetMillis()).
+		Set("SeenReplyCount", "Threads.ReplyCount").
```
Comment (lieut-data) [inline store/sqlstore/thread_store.go:614]: Are we missing something to effectively JOIN `ThreadMemberships` and `Threads` together vs. creating a cross join?

File: store/sqlstore/thread_store.go [hunk 15]
```diff
@@ -656,13 +639,20 @@ func (s *SqlThreadStore) MarkAllAsReadByTeam(userId, teamId string) error {
...
+
+	} else {
+		querySQL = s.getQueryBuilder().Update("ThreadMemberships", "Threads")
+	}
+	query, args, err := querySQL.
 		Set("LastViewed", timestamp).
 		Set("UnreadMentions", 0).
 		Set("LastUpdated", model.GetMillis()).
+		Set("SeenReplyCount", sq.Expr("Threads.ReplyCount")).
+		Where(sq.Eq{"ThreadMemberships.PostId": membershipIds}).
```
Comment (lieut-data) [inline store/sqlstore/thread_store.go:654]: (Same as above.)

File: store/sqlstore/thread_store.go [hunk 16]
```diff
@@ -677,29 +667,59 @@ func (s *SqlThreadStore) MarkAllAsReadByTeam(userId, teamId string) error {
...
+		return errors.Wrapf(err, "failed to create SQL query to count seen reply posts for userId=%s, threadId=%s", userId, threadId)
+	}
+	var seenReplyCount int64
+	err = transaction.Get(&seenReplyCount, queryString, args...)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get seen reply count for userId=%s threadId=%s", userId, threadId)
 	}
 
-	_, err = s.GetMasterX().Exec(query, args...)
+	vals := map[string]interface{}{
```
Comment (lieut-data) [inline store/sqlstore/thread_store.go:694]: Any particular reason not to use the query builder?

File: store/sqlstore/thread_store.go (hunk 7 above)
Comment (lieut-data) [reply store/sqlstore/thread_store.go:683]: If we inline things, can we do it all in a `SELECT CASE WHEN ...`?

File: store/sqlstore/thread_store.go [hunk 17]
```diff
@@ -1010,12 +1061,12 @@ func (s *SqlThreadStore) DeleteOrphanedRows(limit int) (deleted int64, err error
 // return number of unread replies for a single thread
 func (s *SqlThreadStore) GetThreadUnreadReplyCount(threadMembership *model.ThreadMembership) (unreadReplies int64, err error) {
 	query, args, err := s.getQueryBuilder().
-		Select("COUNT(Posts.Id)").
-		From("Posts").
+		Select("(Threads.ReplyCount - ThreadMemberships.SeenReplyCount)").
+		From("ThreadMemberships").
+		InnerJoin("Threads ON ThreadMemberships.PostId = Threads.PostId").
```
Comment (lieut-data) [inline store/sqlstore/thread_store.go:1066]: Will an `INNER JOIN` result in a `ErrNotFound` error if for some reason the corresponding `Threads` row doesn't exist? (vs. `0` before)

File: store/sqlstore/thread_store.go [hunk 18]
```diff
@@ -1028,3 +1079,21 @@ func (s *SqlThreadStore) GetThreadUnreadReplyCount(threadMembership *model.Threa
...
+			sq.Eq{"Posts.DeleteAt": 0},
+			sq.LtOrEq{"Posts.CreateAt": timestamp},
+		}).
+		ToSql()
+	if err != nil {
+		return 0, errors.Wrapf(err, "failed to create SQL query to count reply posts in threadId=%s until timestamp=%d", threadId, timestamp)
+	}
+	var seenReplyCount int64
+	err = s.GetReplicaX().Get(&seenReplyCount, queryString, args...)
+	return seenReplyCount, errors.Wrapf(err, "failed to count reply posts in threadId=%s until timestamp=%d", threadId, timestamp)
```
Comment (lieut-data) [inline store/sqlstore/thread_store.go:1098]: nit: while condensed, I find this harder to read than an explicit:
```go
if err != nil {
    return 0, errors.Wrapf(...)
}

return seenReplyCount, nil
```

File: store/sqlstore/thread_store.go [hunk 19]
```diff
@@ -1028,3 +1079,21 @@ func (s *SqlThreadStore) GetThreadUnreadReplyCount(threadMembership *model.Threa
 
 	return
 }
+
+func (s *SqlThreadStore) GetReplyCountUntil(threadId string, timestamp int64) (int64, error) {
```
Comment (lieut-data) [inline store/sqlstore/thread_store.go:1083]: Does this need to be exposed? (and unit tested?)

File: store/storetest/thread_store.go [hunk 20]
```diff
@@ -332,8 +332,8 @@ func testThreadStorePopulation(t *testing.T, ss store.Store) {
 		require.NoError(t, err)
```
Comment (lieut-data) [inline store/storetest/thread_store.go:332]: Are there any new tests we might want to consider?

**Review by lieut-data — CHANGES_REQUESTED** [review]
This has the potential of once again improving performance! Nice work @ashishbhate :)

A few comments below -- looking forward to loadtest results.

**Comment by ashishbhate** [comment]
@lieut-data @agnivade Thanks for the first pass of reviews. As I mentioned before, this PR needs polish and many more tests. I'm targeting this PR to go in for v7.3. Hopefully by then GA would have stablized and I'll have a loadtest instance running :)

**Comment by isacikgoz** [comment]
Big thanks for your contribution! There hasn't been recent activity on this PR so I've marked it as inactive. Following our [inactive contribution process](https://developers.mattermost.com/contribute/getting-started/inactive-contributions/), the PR is now closed due to 30 days or more of inactivity.

However, if you're still interested working on the changes, let us know! Note that we [merged multiple repositories into one monorepo](https://github.com/mattermost/mattermost-server/issues/22420), which may require you to resubmit the changes. 

If you have any questions, don't hesitate to let us know, we're happy to help.
